### PR TITLE
Improve WordPress.com account handling

### LINKED ISSUE #1972: Once Jetpack authorization is invalidated, it may continue to stuck on the Settings page
### Describe the bug:

Once Jetpack authorization is invalid, it may continue to stuck on the settings page.

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/1638a818-c9d4-47be-a9c5-f15dc3a707da

### Steps to reproduce:

1. Make a Jetpack connected and authorized site temporarily inaccessible
2. Go to https://cloud.jetpack.com/ and select the site. Wait for a while. Reconnect Jetpack -> Disconnect Jetpack

   https://github.com/woocommerce/google-listings-and-ads/assets/17420811/94d79dfd-bf4f-4f4b-a202-c1c7f3f2af42
3. Go back to a GLA admin page.
4. GLA redirects to the Settings page and is continuously stuck in loading status.

### Expected behavior:

GLA should redirect to the Jetpack (WordPress.com) reconnection page.

image

### Actual behavior:

Users are unable to reconnect their sites with Jetpack.

### Additional details:

This probably only happens when the authorization is invalidated from Jetpack side. Disconnecting from GLA won't fall into this issue.

Related codes:
- [Set Jeppack connected flag to false when getting 401 from WCS](https://github.com/woocommerce/google-listings-and-ads/blob/2.4.6/src/Internal/DependencyManagement/GoogleServiceProvider.php#L209-L232)
- [`jetpack/connected` API doesn't check the above flag](https://github.com/woocommerce/google-listings-and-ads/blob/2.4.6/src/API/Site/Controllers/Jetpack/AccountController.php#L170-L198), so the `active` value in API response is still `'yes'`
- [Redirect to the corresponding reconnection page when getting 401](https://github.com/woocommerce/google-listings-and-ads/blob/2.4.6/js/src/data/index.js#L29-L45)
- [Redirect to the Setting page when `active` is `'yes'`](https://github.com/woocommerce/google-listings-and-ads/blob/2.4.6/js/src/settings/reconnect-wpcom-account/reconnect-wpcom-account.js#L26-L32)

### LINKED ISSUE #2118: Disconnect Jetpack doesn't work anymore in Connection Test
### Describe the bug:

Before, when you went to the Connection Test page and clicked on Disconnect Jetpack. It was fully disconnected. Now it shows the next error:

`Cannot disconnect Jetpack connection as there are other plugins using it: jetpack, google-listings-and-ads, woocommerce
Please disconnect the connection using My Jetpack.`

Screenshot 2023-09-29 at 15 50 01

Note: This is because now WooCommerce is also connected to Jetpack. 

### Steps to reproduce:

1. Connect GLA 
2. Go to  Connection Test `/wp-admin/admin.php?page=connection-test-admin-page`
3. Click on Dicsonnect Jetpack

### Expected behavior:
Jetpack is fully disconnected

### Actual behavior:
A message error happens and jetpack is not disconnected

### LINKED ISSUE #2284: No option to reconnect to WordPress.com after Jetpack disconnection
### Describe the bug:

We rely on the Jetpack connection package to establish a connection between Google Listings & Ads and WordPress.com. While we do not offer an option to disconnect Jetpack, Jetpack can get disconnected in the following ways:

* Connection issue - something hindering/breaking the Jetpack connection
* Jetpack can be disconnected using the Jetpack debugger (Internal tool: Happiness uses this tool if a customer wants to change the connected JP account)
* Jetpack can be disconnected by the user from WordPress.com

When Jetpack gets disconnected in any of the above ways, we do not offer the option to reconnect Jetpack. Instead, users are presented with a blank screen and the error `There was an error loading Google account info. Please reconnect Jetpack`

### Steps to reproduce - approach 1:

1. Setup Google Listings & Ads and all required connections without Jetpack plugin
2. Go to the JP debugger Pc9OEs-v-p2 and disconnect the Jetpack connection with the tool or disconnect the Jetpack connection from WordPress.com
3. Go to Marketing > Google Listings & Ads. Notice the error "Please add your phone number" and "Please add your store address" is returned in the contact information section. (I notice this error is removed after a while and a blank screen is shown as in "approach 2" below)
4. Notice no option to reconnect Jetpack is shown
5. Go to the test connection page and disconnect MC by clicking "MC Disconnect". 
6. Return to Marketing > Google Listings & Ads notice the following error is returned:

7. Return to the test connection page and attempt to `disconnect Google account` and notice the response "Please reconnect your Jetpack account." is returned. (If this disconnection is possible, it would show the first step of the setup screen but still won't show the option to reconnect to WordPress.com. See approach 2 below)

### Steps to reproduce - approach 2:

1. Setup Google Listings & Ads and all required connections without Jetpack plugin
2. Go to Marketing > Google Listings & Ad > Settings and click Disconnect from all accounts (This removes GMC and MC connection as expected)
3. Go to the Jetpack debugger and disconnect the Jetpack connection with the tool or disconnect the Jetpack connection from WordPress.com
4. Go to Marketing > Google Listings & Ads. Notice the blank screen with a "spinning circle" and the error `There was an error loading Google account info. Please reconnect Jetpack`:

7. Notice there's no way to reconnect to WordPress.com

### Expected behavior:

The option to reconnect to WordPress.com should be shown when the Jetpack connection is disconnected or broken. 

### Actual behavior:

The option to reconnect to WordPress.com isn't shown. 

### Additional info

Based on my test, the only way to reconnect Jetpack is to install the Jetpack plugin:
- On activation, it shows a prompt to fix the broken Jetpack connection. After fixing it, you can reconnect to WordPress.com via GL&A.
- Even if the above prompt isn't shown - initiating a JP connection through the Jetpack plugin would offer the option to reconnect to WordPress.con in GL&A

### LINKED ISSUE #2285: Defining delete plugin data constant triggers a fatal error
### Describe the bug:

When the Jetpack connection is broken or disconnected, defining a constant to delete the plugin's data triggers a fatal error.

### Steps to reproduce:

1. Disconnect Jetpack through any of the approaches in this issue: https://github.com/woocommerce/google-listings-and-ads/issues/2284
2. Define a constant to delete the plugin's data by adding the following code snippet to wp-config.php: 
`define( 'WC_GLA_REMOVE_ALL_DATA', true );`
3. Go to Marketing > Google Listings & Ads and notice the following error:

4. Check the site's recent fatal error log in WooCommerce > Status > Logs and notice the following error:

`2024-02-28T14:21:21+00:00 Critical Uncaught Automattic\WooCommerce\GoogleListingsAndAds\Exception\AccountReconnect: Please reconnect your Jetpack account. in /home/sundayistesting/public_html/wp-content/plugins/google-listings-and-ads/src/Exception/AccountReconnect.php:26 CONTEXT: {"error":{"type":1,"file":"\/home\/sundayistesting\/public_html\/wp-content\/plugins\/google-listings-and-ads\/src\/Exception\/AccountReconnect.php","line":26},"backtrace":["","#0 \/home\/sundayistesting\/public_html\/wp-content\/plugins\/google-listings-and-ads\/src\/Internal\/DependencyManagement\/GoogleServiceProvider.php(237): Automattic\\WooCommerce\\GoogleListingsAndAds\\Exception\\AccountReconnect::jetpack_disconnected()","#1 \/home\/sundayistesting\/public_html\/wp-content\/plugins\/google-listings-and-ads\/src\/Internal\/DependencyManagement\/GoogleServiceProvider.php(209): Automattic\\WooCommerce\\GoogleListingsAndAds\\Internal\\DependencyManagement\\GoogleServiceProvider->handle_unauthorized_error(Object(Automattic\\WooCommerce\\GoogleListingsAndAds\\Vendor\\GuzzleHttp\\Psr7\\Request), Object(Automattic\\WooCommerce\\GoogleListingsAndAds\\Vendor\\GuzzleHttp\\Psr7\\Response))","#2 \/home\/sundayistesting\/public_html\/wp-content\/plugins\/google-listings-and-ads\/vendor\/guzzlehttp\/promises\/src"]}`

### Expected behavior:

Adding a constant to completely delete the plugin's data shouldn't trigger any fatal error.

### Actual behavior:

It's triggering a fatal error.

### Support interactions

* https://wordpress.org/support/topic/unlink-wordpress-account/#post-17207809
* 7750761-zd-a8c

### Additional information

It's possible there are other ways to trigger the fatal error. However, this is the method I discovered to trigger it.

### PR DESCRIPTION
**Pull request by mikkamp**
### Changes proposed in this Pull Request:

This PR is intended to improve the handling of the WordPress.com connection status. There are several fixes within this PR that handled various scenarios.

#### Uncaught Fatal Error
When we reset all the plugin data using the snippet `define( 'WC_GLA_REMOVE_ALL_DATA', true );`
And at the same time disconnect the WordPress.com connection externally. We end up with a scenario where it assumes the connection is still intact but it hasn't accepted the TOS conditions yet. This triggers a request to WCS which fails because the token is no longer valid, and in turn triggers a ReconnectAccount exception. In this PR we make sure this exception is caught and handled.

Closes #2285 

#### Disconnect during onboarding
If we trigger an external disconnect during onboarding, we end up with a situation where the local token in the DB is still valid but once an external request is sent it returns a response which triggers a ReconnectAccount exception.
The UI tries to resolve this by checking if the WordPress.com account is still connected. Since this only checks the local token it assumes it's connected.

In this PR we resolve that scenario by doing a more extensive check with the token. We both confirm it's available and send a request to a token endpoint to validate the token hasn't been invalidated externally. This ensures the UI gets a response that the account is no longer connected and can once again show the connect button.

Closes #2284

#### Disconnected after onboarding
If we trigger an external disconnect after onboarding, we end up with the same situation where the local token in the DB is still valid but once an external request is sent it returns a response which triggers a ReconnectAccount exception.

In this PR we again resolve that both by doing an external request to validate the token, and by updating the option `JETPACK_CONNECTED` to reflect the actual status. This allows the UI to catch the error and stay on the reconnect page without assuming it's connected.

Closes #1972

#### Remove disconnect button
This PR also removes the disconnect button on the Connection Test page, as it can only perform a "soft connect", which will always show WooCommerce is connected so we are unable to disconnect.
The intention is to handle this better within WooCommerce or some other higher level, at the moment the Jetpack plugin can be installed to disconnect fully.

Closes #2118

#### Jetpack packages update
To ensure we are working with the latest version of the Jetpack connect I also updated the composer packages. According to the changelogs there are mostly minor fixes. The major version bump is for dropping PHP < 7. Since we have a minimum of PHP 7.4 those changes don't affect anything.
`composer update automattic/jetpack* --with-all-dependencies`
```
- Upgrading automattic/jetpack-a8c-mc-stats (v1.4.22 => v2.0.0)
- Upgrading automattic/jetpack-admin-ui (v0.2.24 => v0.3.2)
- Upgrading automattic/jetpack-autoloader (v2.12.0 => v3.0.2)
- Upgrading automattic/jetpack-config (v1.15.4 => v2.0.0)
- Upgrading automattic/jetpack-connection (v1.59.0 => v2.3.2)
- Upgrading automattic/jetpack-constants (v1.6.23 => v2.0.0)
- Upgrading automattic/jetpack-redirect (v1.7.27 => v2.0.0)
- Upgrading automattic/jetpack-roles (v1.4.25 => v2.0.0)
- Upgrading automattic/jetpack-status (v1.18.5 => v2.1.0)
```

#### Uncaught Fatal Error
1. Go through the onboarding steps to connect at least the accounts in step 1
2. Abort onboarding
3. Visit the page /wp-admin/options.php (or directly in the DB) and set the option `gla_wp_tos_accepted` to 0
4. Go to the URL `jptools.wordpress.com/debug/` and find your site
5. Scroll down to the section Blog Details and click Disconnect Jetpack
6. Restart onboarding from the beginning
7. Confirm we are able to connect a WordPress.com account and continue onboarding without issues

#### Disconnect during onboarding
1. Go through the onboarding steps to connect at least the accounts in step 1
2. Abort onboarding
3. Go to the URL `jptools.wordpress.com/debug/` and find your site
4. Scroll down to the section Blog Details and click Disconnect Jetpack
5. Restart onboarding from the beginning
6. Confirm that we see a request to `/wc/gla/jetpack/connected` with a response similar to:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```

#### Disconnect after onboarding
1. Go through all the onboarding steps
2. Go to the URL `jptools.wordpress.com/debug/` and find your site
3. Scroll down to the section Blog Details and click Disconnect Jetpack
4. Refresh one of the GLA pages like Dashboard or Settings
5. Confirm that we are redirected to a reconnect page
image
6. Confirm that we see a request to `/wc/gla/jetpack/connected` with a response similar to:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```
7. Try the connect button on the Reconnect Page and confirm that all the pages work again after this is successful.

#### Test Connection Test page
1. Simulate the same scenario as one of the previous where we disconnect the WordPress.com account externally
2. Visit the Connection Test page and expect to see a section with a reconnect button (similar as is done through the UI). Reconnect button should only be visible in cases where we still have a local connection token, but the token is no longer valid.
image
3. Try the status button and expect a response like:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```
4. Try the reconnect button and confirm that we can correctly connect again (same as in the UI)

### Changelog entry
* Fix - Improve WordPress.com account handling.

### CODE REVIEW

File: src/API/Site/Controllers/Jetpack/AccountController.php [hunk 1]
```diff
@@ -31,6 +31,13 @@ class AccountController extends BaseOptionsController {
 	 */
 	protected $middleware;
 
+	/**
+	 * Retain the connected state to prevent multiple external calls to validate the token.
+	 *
+	 * @var bool
+	 */
+	private $connected;
```
Comment (puntope) [inline src/API/Site/Controllers/Jetpack/AccountController.php]: 💅  WDYT to rename this to `$is_jetpack_connected_state`?  As we are storing the `is_jetpack_connected()` result

**Comment by puntope** [comment]
Doesn't seem to be related to this PR as it's happening with all the options. When I set `gla_wp_tos_accepted` as 0 I got a fatal, However, the option got updated. It may be my config regarding COT?

```shell
/html/wp-content/plugins/wcs-connect.php on line 120PHP message: PHP Fatal error:  Uncaught Exception: The authoritative table for orders storage can't be changed while there are orders out of sync in /var/www/html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:313
jt-test-nginx-1      | Stack trace:
jt-test-nginx-1      | #0 [internal function]: Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController->process_pre_update_option()
jt-test-nginx-1      | #1 /var/www/html/wp-content/plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php(158): call_user_func_array()
jt-test-nginx-1      | #2 /var/www/html/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController->__call()
jt-test-nginx-1      | #3 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters()
jt-test-nginx-1      | #4 /var/www/html/wp-includes/option.php(784): apply_filters()
jt-test-nginx-1      | #5 /var/www/html/wp-admin/options.php(339): update_option()
jt-test-nginx-1      | #6 {main}
jt-test-nginx-1      |   thrown in /var/www/html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php on line 313PH
```

**Comment by puntope** [comment]
> Go to the URL jptools.wordpress.com/debug/ and find your site
> Scroll down to the section Blog Details and click Disconnect Jetpack

Going into `jptools.wordpress.com/debug/` shows a success message after selecting the site, but I didn't see any "Blog Details" sections. Did I miss some steps? 

Screenshot 2024-03-02 at 17 15 01

**Comment by mikkamp** [comment]
> When I set gla_wp_tos_accepted as 0 I got a fatal, However, the option got updated. It may be my config regarding COT?

I got the same fatal and reported it here: https://github.com/woocommerce/woocommerce/issues/45225

**Comment by puntope** [comment]
> Going into jptools.wordpress.com/debug/ shows a success message after selecting the site, but I didn't see any "Blog Details" sections. Did I miss some steps?

Working already. It was finally the proxy as you mentioned.

**Comment by puntope** [comment]
> Disconnect during onboarding

In this step, I'm not sure if the results are as expected. 

When I go back to Onboarding after disconnect. I get an error message. Wouldn't be better to go to Step 1?

https://github.com/woocommerce/google-listings-and-ads/assets/5908855/40387eaf-38ec-4607-996e-835b6dd1cc76

Also the response I got was:

```js
{active: "no", owner: "no", displayName: "", email: ""}
```

**Comment by puntope** [comment]
> Test Connection Test page

Not sure if It's expected but "Reconnect to Worpress.com" appears in Connection Page even when I'm connected.

**Comment by puntope** [comment]
Hi @mikkamp, thanks for your help with my Jetpack Debug problems as well as for reporting the issue with the fatal error.

I tested all the scenarios, only one I have doubts is in the _Disconnect during Onboarding_ case

Please see my comments here https://github.com/woocommerce/google-listings-and-ads/pull/2289#issuecomment-1976807791

**Comment by mikkamp** [comment]
> When I go back to Onboarding after disconnect. I get an error message. Wouldn't be better to go to Step 1?

That is an excellent point and I wholeheartedly agree. This PR focuses on handling the status from the backend, so the UI can have a better endpoint to rely on it's connection status. Since your suggestion would be UI related, I'm logging it in a separate issue here: https://github.com/woocommerce/google-listings-and-ads/issues/2295

Is there anything else besides this issue to followup on? Or a scenario where the endpoint is returning an incorrect status?

You mentioned you got the response:
```js
{
	active: "no",
	owner: "no",
	displayName: "",
	email: ""
}
```

That looks like the status that would be received after the token has been cleaned up locally. Which could be achieved with either the Jetpack plugin disconnecting locally. Or alternatively after initiating a reconnect, based on the changes here we now call:
```php
$result = $this->manager->reconnect();
```

What that reconnect does is it clears up the old tokens, before creating a new connect URL to be redireced to. Is it possible that response was received after doing the reconnect?

**Review by puntope — APPROVED** [review]
THanks for all the explanations @mikkamp  ✅ 

THere is no blocker anymore for me as all the steps worked as described in the description.

**Pull request merged**